flink-libraries/flink-state-processing-api/src/test/java/org/apache/flink/state/api/SavepointDeepCopyTest.java

```diff
@@ -0,0 +1,208 @@
+				"state", Types.TUPLE(Types.STRING, Types.STRING));
+			state = getRuntimeContext().getState(stateDescriptor);
+		}
+
+		@Override
+		public void readKey(
+			String key,
+			Context ctx,
+			Collector<Tuple2<String, String>> out) throws Exception {
+			out.collect(state.value());
+		}
+	}
+
+	/**
+	 * Test savepoint deep copy. This method tests the savepoint deep copy by:
+	 * <ul>
+	 * <li>create {@code savepoint1} with operator {@code Operator1}, make sure it has more state files in addition to
+	 * _metadata
+	 * <li>create {@code savepoint2} from {@code savepoint1} by adding a new operator {@code Operator2}
+	 * <li>check all state files in {@code savepoint1}'s directory are copied over to {@code savepoint2}'s directory
+	 * <li>read the state of {@code Operator1} from {@code savepoint2} and make sure the number of the keys remain same
+	 * </ul>
+	 * @throws Exception throw exceptions when anything goes wrong
+	 */
+
+	@Test
+	public void testSavepointDeepCopy() throws Exception {
+
+        // set up the execution environment
+		ExecutionEnvironment env = ExecutionEnvironment.getExecutionEnvironment();
+
+		// construct DataSet
+		DataSet<String> words = env.fromElements(TEXT.split(" "));
+
+		// create BootstrapTransformation
+		BootstrapTransformation<String> transformation = OperatorTransformation
+			.bootstrapWith(words)
+			.keyBy(e -> e)
+			.transform(new WordMapBootstrapper());
+
+		File savepointUrl1 = createAndRegisterTempFile(new AbstractID().toHexString());
+		String savepointPath1 = savepointUrl1.getPath();
+
+		// create a savepoint with BootstrapTransformations (one per operator)
+		// write the created savepoint to a given path
+		Savepoint.create(backend, 128)
+			.withOperator("Operator1", transformation)
+			.write(savepointPath1);
+
+		env.execute("bootstrap savepoint1");
+
+		Assert.assertTrue(
+			"Failed to bootstrap savepoint1 with additional state files",
+			Files.list(Paths.get(savepointPath1)).count() > 1
+		);
+
+		Set<String> stateFiles1 = Files.list(Paths.get(savepointPath1))
+			.map(path -> path.getFileName().toString())
+			.collect(Collectors.toSet());
+
+		// create savepoint2 from savepoint1 created above
+		File savepointUrl2 = createAndRegisterTempFile(new AbstractID().toHexString());
+		String savepointPath2 = savepointUrl2.getPath();
+
+		ExistingSavepoint savepoint2 = Savepoint.load(env, savepointPath1, backend);
+		savepoint2
+			.withOperator("Operator2", transformation)
+			.write(savepointPath2);
+		env.execute("create savepoint2");
+
+		Assert.assertTrue("Failed to create savepoint2 from savepoint1 with additional state files",
+			Files.list(Paths.get(savepointPath2)).count() > 1
+		);
+
+		Set<String> stateFiles2 = Files.list(Paths.get(savepointPath2))
+			.map(path -> path.getFileName().toString())
+			.collect(Collectors.toSet());
+
+		// check every state file in savepoint1 are also in savepoint2
```

Comment:
If you're adding a comment to an Assert then it should be in the error message.

Reply:
Good idea.